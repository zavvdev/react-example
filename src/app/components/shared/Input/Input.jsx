import PropTypes from "prop-types";
import cx from "clsx";
import { useInputStyles } from "app/components/shared/Input/Input.styles";

export function Input({
  name,
  placeholder,
  className,
  errorText,
  onChange,
  onBlur,
  value,
}) {
  const classes = useInputStyles();
  const inputClasses = cx(
    classes.input,
    {
      [classes.inputError]: !!errorText,
    },
    className,
  );

  return (
    <div>
      <input
        name={name}
        placeholder={placeholder}
        className={inputClasses}
        onChange={onChange}
        onBlur={onBlur}
        value={value}
      />
      {!!errorText && <div className={classes.error}>{errorText}</div>}
    </div>
  );
}

Input.propTypes = {
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  onBlur: PropTypes.func,
  placeholder: PropTypes.string,
  className: PropTypes.string,
  errorText: PropTypes.string,
};

Input.defaultProps = {
  onBlur: undefined,
  placeholder: undefined,
  className: undefined,
  errorText: undefined,
};
