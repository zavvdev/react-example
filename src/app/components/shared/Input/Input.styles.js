import { createUseStyles } from "react-jss";

export const useInputStyles = createUseStyles((theme) => ({
  input: {
    padding: "5px 10px",
    border: `1px solid ${theme.borderColor}`,
    borderRadius: "2px",
    outline: "none",
    boxSizing: "border-box",
    backgroundColor: theme.backgroundColor,
    color: theme.typographyColor,
    fontSize: "15px",
    "&:focus": {
      border: `1px solid ${theme.palette.primary.main}`,
    },
  },
  inputError: {
    border: `1px solid ${theme.palette.red.main}`,
    "&:focus": {
      border: `1px solid ${theme.palette.red.main}`,
    },
    "&::placeholder": {
      color: theme.palette.red.main,
    },
  },
  error: {
    color: theme.palette.red.main,
    fontSize: "11px",
    marginTop: "3px",
  },
}));
