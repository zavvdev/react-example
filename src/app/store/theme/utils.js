import { localStorageService } from "app/services/LocalStorageService";
import { IS_DARK_THEME_LOCAL_STORAGE_KEY } from "app/store/theme/config";

export const saveDarkModeState = (darkModeState) => {
  localStorageService.set(IS_DARK_THEME_LOCAL_STORAGE_KEY, `${darkModeState}`);
};

export const getDarkModeState = () => {
  const rawStorageValue = localStorageService.get(
    IS_DARK_THEME_LOCAL_STORAGE_KEY,
  );
  return !!JSON.parse(rawStorageValue);
};
