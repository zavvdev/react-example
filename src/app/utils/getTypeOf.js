export function getTypeOf(object) {
  const objectType = typeof object;
  let returnType = objectType;
  if (objectType === "object") {
    const stringifiedObject = Object.prototype.toString.call(object);
    returnType = stringifiedObject
      .replace(/^\[object (\S+)\]$/, "$1")
      .toLowerCase();
  }
  return returnType;
}
