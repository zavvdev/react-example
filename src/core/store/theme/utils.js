import { LOCAL_STORAGE_KEYS } from "core/config/localStorage";
import { localStorageService } from "core/services/LocalStorageService";

export const saveDarkModeState = (darkModeState) => {
  localStorageService.set(
    LOCAL_STORAGE_KEYS.isThemeDarkMode,
    `${darkModeState}`,
  );
};

export const getDarkModeState = () => {
  const rawStorageValue = localStorageService.get(
    LOCAL_STORAGE_KEYS.isThemeDarkMode,
  );
  return !!JSON.parse(rawStorageValue);
};
