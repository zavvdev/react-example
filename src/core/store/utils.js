import { all, call, spawn } from "redux-saga/effects";
import { errorTrackingService } from "core/services/ErrorTrackingService";

export const combineMiddleware = (middlewareList) => function* combiner() {
  const getRootGenerator = (middleware) => function* rootGenerator() {
    while (true) {
      try {
        yield call(middleware);
        break;
      } catch (e) {
        errorTrackingService.reportError(e);
      }
    }
  };

  yield all(middlewareList.map((middleware) => {
    return spawn(getRootGenerator(middleware));
  }));
};
