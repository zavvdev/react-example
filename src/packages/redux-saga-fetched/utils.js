import { ACTION_TYPES, DATA_STATUS_TYPES } from "packages/redux-saga-fetched/config";

export const createActionType = ({ createdKey, actionTypePattern }) => {
  return `${createdKey}_${actionTypePattern}`;
};

export const createKey = (key) => key.join("_");

export const createActionTypePatterns = (domain) => ({
  request: `${domain}@${ACTION_TYPES.request}`,
  success: `${domain}@${ACTION_TYPES.success}`,
  failure: `${domain}@${ACTION_TYPES.failure}`,
  invalidate: `${domain}@${ACTION_TYPES.invalidate}`,
  reset: `${domain}@${ACTION_TYPES.reset}`,
});

export const createState = ({
  isLoading,
  isFetching,
  isLoaded,
  isError,
  isValid,
  status,
  data,
}) => ({
  isLoading,
  isFetching,
  isLoaded,
  isError,
  isValid,
  status,
  data,
});

export const createResetState = () => createState({
  isLoading: false,
  isFetching: false,
  isLoaded: false,
  isError: false,
  isValid: false,
  status: DATA_STATUS_TYPES.reset,
  data: null,
});

export const createRequestState = ({ state, payload }) => {
  const isLoading = !state[payload.createdKey]?.data;
  const isFetching = !!state[payload.createdKey]?.data;
  return createState({
    isLoading,
    isFetching,
    isLoaded: false,
    isError: false,
    isValid: false,
    status: isLoading ? DATA_STATUS_TYPES.loading : DATA_STATUS_TYPES.fetching,
    data: state[payload.createdKey]?.data || null,
  });
};

export const createSuccessState = ({ payload }) => createState({
  isLoading: false,
  isFetching: false,
  isLoaded: true,
  isError: false,
  isValid: true,
  status: DATA_STATUS_TYPES.loaded,
  data: payload.data,
});

export const createFailureState = ({ state, payload }) => createState({
  isLoading: false,
  isFetching: false,
  isLoaded: false,
  isError: true,
  isValid: false,
  status: DATA_STATUS_TYPES.error,
  data: state[payload.createdKey]?.data || null,
});

export const createInvalidateState = ({ state, payload }) => createState({
  isLoading: state[payload.createdKey]?.isLoading,
  isFetching: state[payload.createdKey]?.isFetching,
  isLoaded: state[payload.createdKey]?.isLoaded,
  isError: state[payload.createdKey]?.isError,
  isValid: false,
  status: state[payload.createdKey]?.status,
  data: state[payload.createdKey]?.data,
});
