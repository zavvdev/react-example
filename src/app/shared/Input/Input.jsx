import PropTypes from "prop-types";
import cx from "clsx";
import { useInputStyles } from "app/shared/Input/Input.styles";

export function Input({
  type,
  name,
  placeholder,
  className,
  errorText,
  onChange,
  onBlur,
  value,
}) {
  const classes = useInputStyles();
  const inputClasses = cx(
    classes.input,
    {
      [classes.inputError]: Boolean(errorText),
    },
    className,
  );

  return (
    <div>
      <input
        type={type}
        name={name}
        placeholder={placeholder}
        className={inputClasses}
        onChange={onChange}
        onBlur={onBlur}
        value={value}
      />
      {Boolean(errorText) && <div className={classes.error}>{errorText}</div>}
    </div>
  );
}

Input.propTypes = {
  type: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  onBlur: PropTypes.func,
  placeholder: PropTypes.string,
  className: PropTypes.string,
  errorText: PropTypes.string,
};

Input.defaultProps = {
  onBlur: undefined,
  placeholder: undefined,
  className: undefined,
  errorText: undefined,
};
