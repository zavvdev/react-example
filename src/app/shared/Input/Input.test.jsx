import { fireEvent, screen } from "@testing-library/react";
import { Input } from "app/shared/Input/Input";
import { testUtils } from "__tests__/utils";

const properties = {
  type: "text",
  name: "mock",
  placeholder: "Mock",
  className: "mock",
  errorText: "error mock",
  onChange: jest.fn(),
  onBlur: jest.fn(),
  value: "42",
};

test("should render Input", () => {
  const result = testUtils.render(
    <Input
      type={properties.type}
      name={properties.name}
      onChange={properties.onChange}
      value={properties.value}
    />,
  );
  expect(result.asFragment()).toMatchSnapshot();
});

test("should have a placeholder", () => {
  testUtils.render(
    <Input
      type={properties.type}
      name={properties.name}
      onChange={properties.onChange}
      value={properties.value}
      placeholder={properties.placeholder}
    />,
  );
  expect(
    screen.getByPlaceholderText(properties.placeholder),
  ).toBeInTheDocument();
});

test("should have a custom className", () => {
  const result = testUtils.render(
    <Input
      type={properties.type}
      name={properties.name}
      onChange={properties.onChange}
      value={properties.value}
      className={properties.className}
    />,
  );
  expect(
    result.container.querySelectorAll(`.${properties.className}`).length,
  ).toBe(1);
});

test("should have error text", () => {
  testUtils.render(
    <Input
      type={properties.type}
      name={properties.name}
      onChange={properties.onChange}
      value={properties.value}
      errorText={properties.errorText}
    />,
  );
  expect(screen.getByText(properties.errorText)).toBeInTheDocument();
});

test("should call onChange function", () => {
  testUtils.render(
    <Input
      id="mock"
      type={properties.type}
      name={properties.name}
      onChange={properties.onChange}
      value={properties.value}
      placeholder={properties.placeholder}
    />,
  );
  fireEvent.change(screen.getByPlaceholderText(properties.placeholder), {
    target: { value: "123" },
  });
  expect(properties.onChange).toBeCalled();
});

test("should call onBlur function", () => {
  testUtils.render(
    <Input
      id="mock"
      type={properties.type}
      name={properties.name}
      onChange={properties.onChange}
      value={properties.value}
      placeholder={properties.placeholder}
      onBlur={properties.onBlur}
    />,
  );
  fireEvent.blur(screen.getByPlaceholderText(properties.placeholder));
  expect(properties.onBlur).toBeCalled();
});
