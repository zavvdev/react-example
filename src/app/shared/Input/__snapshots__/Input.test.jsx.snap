// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Input 1`] = `
<DocumentFragment>
  <div>
    <input
      class="input"
      name="mock"
      type="text"
      value="42"
    />
  </div>
</DocumentFragment>
`;
